src/sfctl/custom_cluster.py

```diff
@@ -47,32 +53,84 @@ def select(endpoint, cert=None, key=None, pem=None, ca=None, no_verify=False):
     :param str pem: Path to client certificate, as a .pem file
     :param str ca: Path to CA certs directory to treat as valid or CA bundle
     file
+    :param bool aad: Indicate using Azure Active Directory authentication
     :param bool no_verify: Disable verification for certificates when using
     HTTPS, note: this is an insecure option and should not be used for
     production environments
     """
-    from sfctl.config import (set_ca_cert, set_cert, set_cluster_endpoint,
-                              set_no_verify)
+    from sfctl.config import (set_ca_cert, set_auth,
+                              set_cluster_endpoint, client_endpoint,
+                              set_no_verify, no_verify_setting)
     from msrest import ServiceClient, Configuration
-    from sfctl.auth import ClientCertAuthentication
+    from sfctl.auth import ClientCertAuthentication, AdalAuthentication
 
-    select_arg_verify(endpoint, cert, key, pem, ca, no_verify)
+    select_arg_verify(endpoint, cert, key, pem, ca, aad, no_verify)
 
-    client_cert = None
-    if pem:
-        client_cert = pem
-    elif cert:
-        client_cert = (cert, key)
+    access_token = None
+    if aad:
+        endpoint_backup = client_endpoint()
+        noverify_backup = no_verify_setting()
+        set_cluster_endpoint(endpoint)
+        set_no_verify(no_verify)
 
-    rest_client = ServiceClient(
-        ClientCertAuthentication(client_cert, ca, no_verify),
-        Configuration(endpoint)
-    )
+        access_token = sf_get_aad_token()
 
-    # Make sure basic GET request succeeds
-    rest_client.send(rest_client.get('/')).raise_for_status()
+        set_cluster_endpoint(endpoint_backup)
+        set_no_verify(noverify_backup)
+
+        rest_client = ServiceClient(
+            AdalAuthentication(access_token, no_verify),
+            Configuration(endpoint)
+        )
+
+        # Make sure basic GET request succeeds
+        rest_client.send(rest_client.get('/')).raise_for_status()
+    else:
+        client_cert = None
+        if pem:
+            client_cert = pem
+        elif cert:
+            client_cert = (cert, key)
+
+        rest_client = ServiceClient(
+            ClientCertAuthentication(client_cert, ca, no_verify),
+            Configuration(endpoint)
+        )
+
+        # Make sure basic GET request succeeds
+        rest_client.send(rest_client.get('/')).raise_for_status()
 
-    set_ca_cert(ca)
-    set_cert(pem, cert, key)
-    set_no_verify(no_verify)
     set_cluster_endpoint(endpoint)
+    set_no_verify(no_verify)
+    set_ca_cert(ca)
+    set_auth(pem, cert, key, access_token)
+
+def sf_get_aad_token():
+    """Get AAD token"""
+    from sfctl.config import set_config_value
+    from sfctl.apiclient import create
+    from datetime import datetime
+    from datetime import timedelta
+
+    set_config_value('security', 'none')
+    sf_client = create(None)
```

Comment:
You don't need to create a client here, you can directly call into the python SDK which will allow you to customise the security settings used.